### PR TITLE
feat(Typescript): Adding CDK example chaining S3, SNS, SQS and Lambda

### PR DESCRIPTION
**Pull request by ericglasser**
Fixes [#520](https://github.com/aws-samples/aws-cdk-examples/issues/520)
---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: typescript/s3-sns-lambda-chain/cdk.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    "@aws-cdk/aws-ecr-assets:dockerIgnoreSupport": true,
+    "@aws-cdk/aws-secretsmanager:parseOwnedSecretName": true,
+    "@aws-cdk/aws-kms:defaultKeyPolicies": true,
+    "@aws-cdk/aws-s3:grantWriteWithoutAcl": true,
+    "@aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount": true,
+    "@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
+    "@aws-cdk/aws-efs:defaultEncryptionAtRest": true,
+    "@aws-cdk/aws-lambda:recognizeVersionProps": true,
+    "@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": true
+  }
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/cdk.json:17]:
```suggestion
  "context": {
    "@aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId": true,
    "@aws-cdk/core:stackRelativeExports": true,
    "@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
    "@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": true
  }
```
Removed the flags that are deprecated in v2.
[v2 Feature flag docs](https://docs.aws.amazon.com/cdk/v2/guide/featureflags.html)

File: typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import * as sns from '@aws-cdk/aws-sns';
+import * as s3 from '@aws-cdk/aws-s3';
+import * as lambdaNodejs from '@aws-cdk/aws-lambda-nodejs';
+import * as lambdaEvents from '@aws-cdk/aws-lambda-event-sources';
+import * as snsSubscriptions from '@aws-cdk/aws-sns-subscriptions';
+import * as s3Notifications from '@aws-cdk/aws-s3-notifications';
+import { join } from 'path'
+
+export class S3SnsSqsLambdaChainStack extends cdk.Stack {
+  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts:13]:
```suggestion
import { App, Stack, StackProps, Duration } from 'aws-cdk-lib';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as lambdaNodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as lambdaEvents from 'aws-cdk-lib/aws-lambda-event-sources';
import * as snsSubscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as s3Notifications from 'aws-cdk-lib/aws-s3-notifications';
import { join } from 'path'

export class S3SnsSqsLambdaChainStack extends Stack {
  constructor(scope: App, id: string, props?: StackProps) {
```

Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    super(scope, id, props);
+
+    // Note: A dead-letter queue is optional but it helps capture any failed messages
+    const deadLetterQueue = new sqs.Queue(
+        this,
+        'deadLetterQueueId', {
+          retentionPeriod: cdk.Duration.days(7)
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts:20]:
```suggestion
          retentionPeriod: Duration.days(7)
```

Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    );
+
+    const uploadQueue = new sqs.Queue(
+      this,
+      'sampleQueueId', {
+        visibilityTimeout: cdk.Duration.seconds(30),
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts:27]:
```suggestion
        visibilityTimeout: Duration.seconds(30),
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+              {
+                storageClass: s3.StorageClass.INFREQUENT_ACCESS,
+                transitionAfter: cdk.Duration.days(30)
+              },
+              {
+                storageClass: s3.StorageClass.GLACIER,
+                transitionAfter: cdk.Duration.days(90)
+              }
+            ]
+          }
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts:70]:
```suggestion
            expiration: Duration.days(365),
            transitions: [
              {
                storageClass: s3.StorageClass.INFREQUENT_ACCESS,
                transitionAfter: Duration.days(30)
              },
              {
                storageClass: s3.StorageClass.GLACIER,
                transitionAfter: Duration.days(90)
              }
            ]
          }
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/package.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  "dependencies": {
+    "@aws-cdk/aws-lambda-nodejs": "*",
+    "@aws-cdk/aws-lambda-event-sources": "*",
+    "@aws-cdk/aws-s3": "*",
+    "@aws-cdk/aws-s3-notifications": "*",
+    "@aws-cdk/aws-sns": "*",
+    "@aws-cdk/aws-sns-subscriptions": "*",
+    "@aws-cdk/aws-sqs": "*",
+    "@aws-cdk/core": "*"
+  }
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/package.json:32]:
```suggestion
  "devDependencies": {
    "@types/jest": "^26.0.10",
    "@types/node": "10.17.27",
    "aws-cdk": "*",
    "jest": "^26.4.2",
    "ts-jest": "^26.2.0",
    "ts-node": "^9.0.0",
    "typescript": "~3.9.7"
  },
  "dependencies": {
   	"aws-cdk-lib": "*",
    "constructs": "*"
  }
```

Would you mind [upgrading this to CDK v2](https://docs.aws.amazon.com/cdk/v2/guide/migrating-v2.html#migrating-v2-v1-uppgrade) for consistency with the other projects we have? I've added code suggestions with all the necessary code to upgrade. It's possible I may have missed something though but let me know if you have trouble.

By the way you'll need to update the lockfile once these suggestions are accepted. Run `npm install` and commit the updated lockfile changes.

File: typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { Template, Match } from '@aws-cdk/assertions';
+import * as cdk from '@aws-cdk/core';
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts:2]:
```suggestion
import { Template, Match } from 'aws-cdk-lib/assertions';
import { App } from 'aws-cdk-lib';
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import * as S3SnsSqsLambdaChain from '../lib/s3_sns_sqs_lambda_chain-stack';
+
+test('SNS Topic Created', () => {
+  /*
+  Test for SNS Topic and Subscription: S3 Upload Event Notification
+  */
+
+  const app = new cdk.App();
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts:10]:
```suggestion
  const app = new App();
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+test('SQS Queue Created', () => {
+  /*
+  Test for SQS Queue:
+    - Queue to process uploads
+    - Dead-letter Queue
+  */
+ 
+  const app = new cdk.App();
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts:26]:
```suggestion
  const app = new App();
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  template.resourceCountIs('AWS::SQS::Queue', 2);
+});
+
+test('S3 Bucket Created', () => {
+  /*
+  Test for S3 Bucket:
+    - Upload Bucket
+  */
+ 
+  const app = new cdk.App();
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts:40]:
```suggestion
  const app = new App();
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+test('Lambda Function Created', () => {
+  /*
+  Test for Lambdas created:
+    - Sample Lambda
+    - Bucket Notification Handler (automatically provisioned)
+  */
+ 
+  const app = new cdk.App();
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/test/s3_sns_sqs_lambda_chain.test.ts:55]:
```suggestion
  const app = new App();
```
Upgrade to CDK v2

File: typescript/s3-sns-lambda-chain/bin/s3_sns_sqs_lambda_chain.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env node
+import * as cdk from '@aws-cdk/core';
+import { S3SnsSqsLambdaChainStack } from '../lib/s3_sns_sqs_lambda_chain-stack';
+
+const app = new cdk.App();
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/bin/s3_sns_sqs_lambda_chain.ts:5]:
```suggestion
import { App } from 'aws-cdk-lib';
import { S3SnsSqsLambdaChainStack } from '../lib/s3_sns_sqs_lambda_chain-stack';

const app = new App();
```
Upgrade to CDK v2

**Review by ryparker — CHANGES_REQUESTED** [review]
A few suggestions before we merge. Mostly around upgrading to CDK v2 since we recommend everyone make the leap.

Thanks for contributing 🙏🏻

**Comment by ericglasser** [comment]
@ryparker Thank you for the review I have upgraded to CDK v2, and used your suggestions.

File: typescript/s3-sns-lambda-chain/README.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+3. The SNS Topic has an SQS queue that subscribes to notifications
+4. When the upload event is published, the SNS message is put into the SQS Queue
+5. A Lambda is polling the SQS as its Event Source
+6. The Lambda receives the message from the SQS queue and processes it
+7. If the message is not processed by the Lambda the message eventually ends up in a dead-letter queue
+
+
+## Build and Deploy
+
+The `cdk.json` file tells the CDK Toolkit how to execute your app.
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/README.md:46]:
```suggestion
Ensure aws-cdk is installed and your AWS account/region is [bootstrapped](https://docs.aws.amazon.com/cdk/latest/guide/bootstrapping.html).

	  ```bash
	  $ npm install -g aws-cdk
	  $ cdk bootstrap
	  ```

Then build and deploy.
_You will need to have [Docker](https://docs.docker.com/get-docker/) installed and running._

	  ```bash
	  $ npm run build
	  $ cdk deploy
	  ```
```

Added some additional guidance on how to bootstrap, build, and deploy the app.

File: typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts [hunk 14]
```diff
@@ -0,0 +1,97 @@
+import { App, Stack, StackProps, Duration } from 'aws-cdk-lib';
+import * as sqs from 'aws-cdk-lib/aws-sqs';
+import * as sns from 'aws-cdk-lib/aws-sns';
+import * as s3 from 'aws-cdk-lib/aws-s3';
+import * as lambdaNodejs from 'aws-cdk-lib/aws-lambda-nodejs';
+import * as lambdaEvents from 'aws-cdk-lib/aws-lambda-event-sources';
+import * as snsSubscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
+import * as s3Notifications from 'aws-cdk-lib/aws-s3-notifications';
+import { join } from 'path'
+export class S3SnsSqsLambdaChainStack extends Stack {
```
Comment (ryparker) [inline typescript/s3-sns-lambda-chain/lib/s3_sns_sqs_lambda_chain-stack.ts:10]:
```suggestion

export class S3SnsSqsLambdaChainStack extends Stack {
```

Spacing

**Review by ryparker — DISMISSED** [review]
Hey @ericglasser 👋🏻 

Thanks for making those changes. I have one suggestion on the README and another ⛏️ spacing suggestion.

I did run into an error when deploying this today:

```sh
12:47:34 PM | CREATE_FAILED        | AWS::SQS::Queue                 | deadLetterQueueId
Resource handler returned message: "Received an UnknownHostException when attempting to interact with a service. See cause for the exact endpoint that is failing to resolve. If this is happen
ing on an endpoint that previously worked, there may be a network connectivity issue or your DNS cache could be storing endpoints for too long." (RequestToken: df20a05d-43d5-5948-5ee3-b0a584e
75fd7, HandlerErrorCode: GeneralServiceException)
```

Could you see if you're able to reproduce this on your end?

**Comment by ericglasser** [comment]
Hey @ryparker,

Sorry for the delay I didn't get a notification for the update. I made the suggested changes and tested deploying and destroying the environment I was unable to reproduce the issue though are you still seeing it?

**Comment by ericglasser** [comment]
@ryparker circling back on this, I was wondrering if you ever had a chance to review the changes?

**Review by ryparker — APPROVED** [review]

**Review by kaizio — COMMENTED** [review]
Fixed tests

**Review by kaizio — APPROVED** [review]

**Pull request merged**